### PR TITLE
pypeit_collate_1d and HISTORY entries

### PR DESCRIPTION
**Pull request by badpandabear**
This is the first pull request for the pypeit_collate_1d tool and for adding HISTORY entries for PypeIt output files.  
Features in this version:
Grouping spectra from spec1d files by ra/dec or by pixel and coadding
Filtering by slit bitmask (requires spec2d file)
Writing metadata to output directory
Creating a an archive directory with files and metadata for KOA
Adds RA/DEC of objects to RA_OBJ and DEC_OBJ in coadded output.
Accepts [coadd1d] parameters in its .collate1d file, or in a separate .coadd1d file with the same prefix as the .collate1d file.

Items planned for future:
fluxing
removing dependence on spec2d files
refactoring code out of scripts/pypeit_collate_1d.py so that they can appear in api docs.
writing signal to noise information to output
writing a QA PDF
adding weather information (although GUIDFWHM has been added to metadata)
Supporting HISTORY for more types of processing/with more  parameters

### CODE REVIEW

File: doc/collate1d.rst
Comment (jhennawi) [inline doc/collate1d.rst:4]: Nice docs!

File: pypeit/history.py
Comment (jhennawi) [inline pypeit/history.py:1]: Very nice!!

File: pypeit/pypeit.py
Comment (jhennawi) [inline pypeit/pypeit.py]: Can you remove this. Have you tried 

from IPython import embed; embed()

A much more fun way to debug in my opinion;)

File: pypeit/spectrographs/util.py
Comment (jhennawi) [inline pypeit/spectrographs/util.py:46]: Nice, I like this option.

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [inline pypeit/scripts/collate_1d.py]: Do we really want to make the name KOA specific? We support non-Keck instruments as well, which this should also work with. We can adopt the KOA framework, but maybe name it something different.

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [inline pypeit/scripts/collate_1d.py:386]: The default unit should be in arcseconds as we discussed. Maybe I'm missing how it is coded.

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [inline pypeit/scripts/collate_1d.py]: My preference for this argument is that it would not be an astropy.coordinates.Angle object but rather either be a float, or a astropy quantity with units of angle, i.e. u.arcsec or u.deg. But I guess that depends on whether users will be calling this function directly. My point is that there are slightly more lines of code required to instantiate an astropy angle object, as compared to just passing in the units of the matching tolerance.

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [inline pypeit/scripts/collate_1d.py:11]: So far we have had all our scripts import methods and classes from core or other directories, rather than in this module where the module in scripts also has the methods and classes it needs. I actually prefer your approach, but it depends on whether these methods are used elsewhere.

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [inline pypeit/scripts/collate_1d.py]: Should this be in pypeit/utils.py?

**Review by jhennawi — CHANGES_REQUESTED** [review]
Thanks for this great PR @badpandabear. My only substantive comment is how collate1d interfaces with pypeit_flux_setup, which does some similar operations. I'm wondering whether we can use some of this nice code that you have developed to improve pypeit_flux_setup and thus have less code duplicating operations.

File: pypeit/coadd1d.py
Comment (profxj) [inline pypeit/coadd1d.py:124]: very convenient!

File: pypeit/pypeit.py
Comment (badpandabear) [reply pypeit/pypeit.py]: FIxed. Sorry about that, I have a bad habit of putting these in code and forgetting to remove them. Thanks for finding this.

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py:386]: This code is getting the RA/DEC from the SpecObj which is in decimal degrees.  The tolerance does default to arcseconds and is handled in main() at around line 777.

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py]: I changed it to take a floating point value and an optional "unit" argument that defaults to arcseconds.

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py:11]: I was planning to split up pypeit_collate_1d.py in the next pull release. Do you want me to do that in this one?

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py]: I agree, and changed it to ArchiveDir.

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py]: Done

File: pypeit/scripts/collate_1d.py
Comment (profxj) [reply pypeit/scripts/collate_1d.py:11]: I am fine waiting for the next PR

File: pypeit/scripts/collate_1d.py
Comment (profxj) [inline pypeit/scripts/collate_1d.py]: What if the spectrum was only extracted with BOXCAR?
Can we allow for that?

File: pypeit/scripts/collate_1d.py
Comment (profxj) [inline pypeit/scripts/collate_1d.py:507]: Probably should go in `io.py`  (eventually)

File: pypeit/scripts/collate_1d.py
Comment (profxj) [inline pypeit/scripts/collate_1d.py:606]: This is fine, but let me suggest that for the next PR that we
add `spec2d_filename` to the `SpecObj` data model.

Will help us maintain provenance.

File: pypeit/tests/test_history.py
Comment (profxj) [inline pypeit/tests/test_history.py:1]: Thanks for the unit tests!

**Review by profxj — APPROVED** [review]
Great stuff.  Put in a number of comments
but am approving as none of them were 
especially onerous.

File: pypeit/scripts/collate_1d.py
Comment (badpandabear) [reply pypeit/scripts/collate_1d.py]: I added BOX_COUNTS to this check, so that only objects with no counts are excluded. With this change BOXCAR only can be done by changing ex_value in the coadd1d parameters. But it's still not possible to mix BOX and OPT in the same run of collate_pypeit_1d. Would we ever want to do that?

File: pypeit/scripts/collate_1d.py
Comment (jhennawi) [reply pypeit/scripts/collate_1d.py:11]: Fine with me.

File: pypeit/scripts/collate_1d.py
Comment (profxj) [reply pypeit/scripts/collate_1d.py]: good question.  I'll say no for now. :)

**Comment by jhennawi** [comment]
@badpandabear I'm ready to approve here after develop is merged in and tests are posted. I'd suggest that you probably wait for mosfire_oct20 though, since that is next in the merge order.

**Comment by badpandabear** [comment]
Screenshot_unit_test_deimos_collate_1d
Screenshot_devsuite_deimos_collate_1d
Unit tests and devsuite complete

**Review by jhennawi — APPROVED** [review]

**Comment by profxj** [comment]
merging

**Pull request merged**
